lib/foreman_inventory_upload/generators/fact_helpers.rb

```diff
@@ -57,55 +57,141 @@ def cloud_provider(host)
+            # If the fact contains invalid JSON, default to an empty array to prevent upload failures.
+            parsed_insights_array = []
+          end
+
+          parsed_insights_item = parsed_insights_array.find { |item| item['original'] == host.fqdn }
+          # If no matching obfuscated hostname is found in the fact, or if the fact itself was empty,
+          # then dynamically obfuscate the host's FQDN using the `obfuscate_fqdn` helper method.
+          parsed_insights_item&.[]('obfuscated') || obfuscate_fqdn(host.fqdn)
+        else
+          # If hostname obfuscation is not enabled for this host, return the host's original FQDN.
+          host.fqdn
+        end
       end
 
       def obfuscate_fqdn(fqdn)
         "#{Digest::SHA1.hexdigest(fqdn)}.example.com"
       end
 
       def obfuscate_ips?(host)
-        insights_client_setting = fact_value(host, 'insights_client::obfuscate_ip_enabled')
-        insights_client_setting = ActiveModel::Type::Boolean.new.cast(insights_client_setting)
-        return insights_client_setting unless insights_client_setting.nil?
+        # Returns true if IP obfuscation should be applied for a given host.
 
-        Setting[:obfuscate_inventory_ips]
+        # Check the global setting for IP obfuscation.
+        return true if Setting[:obfuscate_inventory_ips]
+
+        # These facts indicate whether IPv4 and IPv6 obfuscation are enabled for this particular host.
+        insights_client_ipv4_setting = fact_value(host, 'insights_client::obfuscate_ipv4_enabled')
+        insights_client_ipv6_setting = fact_value(host, 'insights_client::obfuscate_ipv6_enabled')
+
+        cast_ipv4_setting = ActiveModel::Type::Boolean.new.cast(insights_client_ipv4_setting)
+        cast_ipv6_setting = ActiveModel::Type::Boolean.new.cast(insights_client_ipv6_setting)
+
+        # Determine the final obfuscation decision.
+        # IP obfuscation is enabled if:
+        # - The host's IPv4 obfuscation setting is true OR
+        # - The host's IPv6 obfuscation setting is true OR
+        cast_ipv4_setting || cast_ipv6_setting || false
       end
 
       def host_ips(host)
+        # Determines and returns the IP addresses associated with a host,
+        # applying obfuscation if enabled.
+
+        # If IP obfuscation is enabled for the host
+        # return a representation of obfuscated IP addresses.
         return obfuscated_ips(host) if obfuscate_ips?(host)
 
-        # return a pass through proxy hash in case no obfuscation needed
+        # If IP obfuscation is NOT needed, return a special kind of Hash.
+        # where when you try to access a key in it
+        # if the key doesn't exist, it simply returns the key itself.
+        # This is useful because it means if you try to get an IP from this hash,
+        # you'll just get the original IP back. It allows the calling code to
+        # use the same interface whether obfuscation is applied or not.
         Hash.new { |h, k| k }
```

Comment:
I know experienced Rubyists will call this comment unnecessary, but I vote to keep it in (since I had to refresh my memory there)